### ISSUE TITLE
Clippy fix for unnecessary type cast causes compile error

### ISSUE THREAD

**Issue by gadorlhiac** [rust-lang/rust-clippy#16449]
### Summary

A suggestion for an unnecessary cast on an `f64` results in a compiler error as the full type cannot be inferred.

This was using the `num_traits::pow::Pow` trait and raising an f64 to a fractional exponent. When using rust's `powf` there are no issues with clippy. At any rate, while it resolves correctly, using `pow` was probably a minor bug/not good usage in the actual code, but providing the output here in case it points to something in clippy.

### Reproducer

The clippy output was as follows when trying to run `--fix`:
```
after fixes were automatically applied the compiler reported errors within these files:

  * src/lda.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0282]: type annotations needed
  --> src/lda.rs:84:36
   |
84 |                 &[d_matrix.view(), (alpha.pow(0.5) * l_matrix).view()],
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0282`.
Original diagnostics will follow.

warning: this returns a `Result<_, ()>`
  --> src/lda.rs:39:5
   |
39 |     pub fn solve(&self) -> Result<Array2<f64>, ()> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: use a custom `Error` type instead
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.92.0/index.html#result_unit_err
   = note: `#[warn(clippy::result_unit_err)]` on by default

warning: casting to the same type is unnecessary (`f64` -> `f64`)
  --> src/lda.rs:84:37
   |
84 |                 &[d_matrix.view(), (alpha.pow(0.5) as f64 * l_matrix).view()],
   |                                     ^^^^^^^^^^^^^^^^^^^^^ help: try: `alpha.pow(0.5)`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.92.0/index.html#unnecessary_cast
   = note: `#[warn(clippy::unnecessary_cast)]` on by default

```

The following `main.rs` can reproduce the warning (in addition to others which can be ignored..). Requires `ndarray` and `num_traits`.

```rust
use ndarray::{Array1, Array2, Axis, concatenate};
use num_traits::pow::Pow;

fn solve(
    alpha: f64,
    a_matrix: &Array2<f64>,
    d_matrix: &Array2<f64>,
    l_matrix: &Array2<f64>,
    wls: &Array1<f64>,
) -> Result<Array2<f64>, ()> {
    let (d_augmented, a_augmented) = if alpha != 0.0 {
        (
            &concatenate(
                Axis(0),
                &[d_matrix.view(), (alpha.pow(0.5) as f64 * l_matrix).view()],
            )
            .unwrap(),
            &concatenate(
                Axis(0),
                &[
                    a_matrix.view(),
                    Array2::zeros((l_matrix.shape()[0], wls.shape()[0])).view(),
                ],
            )
            .unwrap(),
        )
    } else {
        (d_matrix, a_matrix)
    };

    Ok(l_matrix.clone())
}

fn main() {
    println!("Hello, world!");
}
```

### Version

```text
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-unknown-linux-gnu
release: 1.92.0
LLVM version: 21.1.3


clippy 0.1.92 (ded5c06cf2 2025-12-08)
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by S1gn3rs** [comment]
Confirmed, I can reproduce as well. I’ll take a look.
@rustbot  claim
